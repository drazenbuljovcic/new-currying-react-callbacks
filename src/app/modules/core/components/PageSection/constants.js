export const DEFAULT_COMPONENT_ID = 'PageSection';
export const DEFAULT_CONTAINER_COMPONENT_ID = 'PageSection-Container';
export const DEFAULT_ITEM_COMPONENT_ID = 'PageSection-Item';