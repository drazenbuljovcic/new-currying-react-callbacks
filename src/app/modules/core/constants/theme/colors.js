const colors = {
  darker: '#652022',
  default: '#d4333e',
  light: '#fb8a49',
  lighter: '#f9c853',
  lightest: '#f6f3d8',

  black: '#000',
  white: '#fff',
};

export default colors;
