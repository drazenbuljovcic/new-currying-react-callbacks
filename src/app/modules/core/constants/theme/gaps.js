const gaps = {
  xxl: '40px',
  xl: '30px',
  large: '25px',
  standard: '20px',
  medium: '15px',
  normal: '10px',
  small: '5px',
  xs: '3px',
};

export default gaps;
