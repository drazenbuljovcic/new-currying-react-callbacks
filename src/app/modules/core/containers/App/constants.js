export const DEFAULT_COMPONENT_ID = 'App';

// header
const HEADER = 'HEADER';
const ITEM = 'ITEM';
const LOGO = 'LOGO';

// main
const MAIN = 'MAIN';
const CONTAINER = 'CONTAINER';
const SECTION = 'SECTION';
const ELEMENT = 'ELEMENT';

// footer
const FOOTER = 'FOOTER';
const CTA = 'CTA';

export const AVAILABLE_CONTAINERS = {
  HEADER,
  MAIN,
  FOOTER,
};

export const AVAILABLE_ITEMS = {
  ITEM,
  LOGO,
  
  CONTAINER,
  SECTION,
  ELEMENT,

  CTA,
};

export const AVAILABLE_COMPONENTS = {
  ...AVAILABLE_CONTAINERS,
  ...AVAILABLE_ITEMS,
};