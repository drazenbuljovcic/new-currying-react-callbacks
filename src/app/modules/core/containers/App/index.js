import React from 'react';

import AppStructureStateProvider, { ACTIVATOR_STATUSES, AVAILABLE_COMPONENTS } from './factory';

import PageSection from '../../components/PageSection';

const {
  HEADER,
  ITEM,
  LOGO,

  MAIN,
  CONTAINER,
} = AVAILABLE_COMPONENTS;

class App extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      AppStructure: AppStructureStateProvider(),
    };

    console.log(this.state.AppStructure.AVAILABLE_COMPONENTS)
  }

  setStatusAs = status => container => (item) => () => {
    if (!ACTIVATOR_STATUSES[status]) {
      throw new Error('Unable to set unknown activator status!');
    }

    const { AppStructure } = this.state;

    if (!AppStructure.getTreeShape().containers.byName[container]) {
      throw new Error('Unable to set status on an unknown container!');
    }

    if (!AppStructure.getTreeShape().items.byName[item]) {
      throw new Error('Unable to set status on an unknown item!');
    }
    
    const component = AppStructure.getTreeStructure()[container][item];
    if (component.compareStatus(status)) return;

    component.toggleStatus({ status });

    this.setState({ AppStructure });
    return AppStructure;
  };

  setActiveStatusForComponent = this.setStatusAs(ACTIVATOR_STATUSES.ON);
  setInactiveStatusForComponent = this.setStatusAs(ACTIVATOR_STATUSES.OFF);

  // containers
  // // header
  setActiveStatusForHeaderChild = this.setActiveStatusForComponent(HEADER);
  setInactiveStatusForHeaderChild = this.setInactiveStatusForComponent(HEADER);
  // // main
  setActiveStatusForMainChild = this.setActiveStatusForComponent(MAIN);
  setInactiveStatusForMainChild = this.setInactiveStatusForComponent(MAIN);

  // items
  // // header
  setActiveStatusForHeaderItem = this.setActiveStatusForHeaderChild(ITEM);
  setInactiveStatusForHeaderItem = this.setInactiveStatusForHeaderChild(ITEM);
  setActiveStatusForHeaderLogo = this.setActiveStatusForHeaderChild(LOGO);
  setInactiveStatusForHeaderLogo = this.setInactiveStatusForHeaderChild(LOGO);

  // // main
  setActiveStatusForMainContainer = this.setActiveStatusForMainChild(CONTAINER);
  setInactiveStatusForMainContainer = this.setInactiveStatusForMainChild(CONTAINER);

  render() {
    const tree = this.state.AppStructure.getTreeStructure();

    return (
      <div className="App">
        <div>
          <button onClick={this.setActiveStatusForHeaderItem}>SET</button>
          <p>{tree[HEADER][ITEM].isActive() ? 'active' : 'inactive'}</p>
          <button onClick={this.setInactiveStatusForHeaderItem}>SET</button>
        </div>

        <div>
          <button onClick={this.setActiveStatusForHeaderLogo}>SET</button>
          <p>{tree[HEADER][LOGO].isActive() ? 'active' : 'inactive'}</p>
          <button onClick={this.setInactiveStatusForHeaderLogo}>SET</button>
        </div>
      </div>
    );
  }
};

export default App;
