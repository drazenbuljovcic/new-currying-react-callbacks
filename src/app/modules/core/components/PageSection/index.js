import React from 'react';
import PropTypes from 'prop-types';

import { COLLECTION_PROP_TYPE } from './prop-types/section.item';

import { DEFAULT_COMPONENT_ID, DEFAULT_CONTAINER_COMPONENT_ID, DEFAULT_ITEM_COMPONENT_ID } from './constants'

import theme from '../../constants/theme';

import DefaultContainer from './Components/Containers/Default';
import DefaultItem from './Components/Items/Default';

import { Wrapper } from './styled.components';

const { colors } = theme;

const PageSection = ({
  id,
  className,
  style,

  items,

  Container,
  ContainerProps,

  Item,
  ItemProps,

  componentId,
}) => {
  const elementProps = {
    id: `${componentId}-${id}`,
    className: `${componentId} ${className}`,
    style,
  };

  const internalContainerProps = {
    ...ContainerProps,
    componentId: DEFAULT_CONTAINER_COMPONENT_ID,
  };

  const internalItemProps = {
    ...ItemProps,
    componentId: DEFAULT_ITEM_COMPONENT_ID,
  };

  return (
    <Wrapper {...elementProps}>
      <Container
        Item={Item}
        ItemProps={internalItemProps}
        items={items}
        {...internalContainerProps} />
    </Wrapper>
  );
};

PageSection.AVAILABLE_CONTAINERS = {
  default: DefaultContainer,
};

PageSection.AVAILABLE_ITEMS = {
  default: DefaultItem,
};

PageSection.AVAILABLE_COLORS = {
  default: colors.default,
  black: colors.default,
};

PageSection.propTypes = {
  items: COLLECTION_PROP_TYPE,

  Container: PropTypes.oneOfType([PropTypes.node, PropTypes.element, PropTypes.func]),
  ContainerProps: PropTypes.shape(),

  Item: PropTypes.oneOfType([PropTypes.node, PropTypes.element, PropTypes.func]),
  ItemProps: PropTypes.shape(),

  componentId: PropTypes.string,
};

PageSection.defaultProps = {
  items: [],

  Container: PageSection.AVAILABLE_CONTAINERS.default,
  ContainerProps: {},

  Item: PageSection.AVAILABLE_ITEMS.default,
  ItemProps: {},

  componentId: DEFAULT_COMPONENT_ID,
};

export default PageSection;
