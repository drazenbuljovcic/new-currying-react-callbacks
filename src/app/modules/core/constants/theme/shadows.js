import colors from './colors';

const shadows = {
  text: {
    normal: `1px 1px 2px ${colors.black}`,
  },
};

export default shadows;
