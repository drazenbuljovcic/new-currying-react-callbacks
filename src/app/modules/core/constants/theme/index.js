import colors from './colors';
import gaps from './gaps';
import shadows from './shadows';

export default {
  colors,
  gaps,
  shadows,
};
