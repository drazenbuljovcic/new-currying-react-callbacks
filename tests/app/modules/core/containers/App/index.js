import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { ThemeProvider } from 'styled-components';
import theme from 'activator/src/app/modules/core/constants/theme';

import PageSection from 'app/modules/core/components/PageSection';

import App, { App as PlainAppComponent } from 'app/modules/core/containers/App';
import AppStructureStateProvider from 'app/modules/core/containers/App/factory';

describe('App component tests', () => {
  const structureProvider = AppStructureStateProvider();

  it('should match it\'s default snapshot', () => {
    const app = shallow(<App />);
    expect(toJson(app)).toMatchSnapshot();
  });
  
  it('should be wrapped in a theme context by default', () => {
    const app = shallow(<App />);
    expect(app.find(ThemeProvider).length).toEqual(1);
  });
  
  it('should be mounted with the factory component tree', () => {
    const themedApp = mount(
      <ThemeProvider theme={theme}>
        <PlainAppComponent />
      </ThemeProvider>
    );

    const appInstance = themedApp.children().instance();
    expect(structureProvider.getTreeShape())
      .toBe(appInstance.state.AppStructure.getTreeShape());;
  });

  it('should render three sections by the default configuration', () => {
    const themedApp = mount(
      <ThemeProvider theme={theme}>
        <PlainAppComponent />
      </ThemeProvider>
    );
    
    expect(themedApp.find(PageSection).length)
      .toEqual(Object.keys(structureProvider.getTreeStructure()).length);
  });
});