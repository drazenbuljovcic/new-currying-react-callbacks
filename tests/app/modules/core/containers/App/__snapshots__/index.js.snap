// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests should match it's default snapshot 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "black": "#000",
        "darker": "#652022",
        "default": "#d4333e",
        "light": "#fb8a49",
        "lighter": "#f9c853",
        "lightest": "#f6f3d8",
        "white": "#fff",
      },
      "gaps": Object {
        "large": "25px",
        "medium": "15px",
        "normal": "10px",
        "small": "5px",
        "standard": "20px",
        "xl": "30px",
        "xs": "3px",
        "xxl": "40px",
      },
      "shadows": Object {
        "text": Object {
          "normal": "1px 1px 2px #000",
        },
      },
    }
  }
>
  <App
    componentId="App"
  />
</ThemeProvider>
`;

exports[`App component tests should match its default snapshot 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "black": "#000",
        "darker": "#652022",
        "default": "#d4333e",
        "light": "#fb8a49",
        "lighter": "#f9c853",
        "lightest": "#f6f3d8",
        "white": "#fff",
      },
      "gaps": Object {
        "large": "25px",
        "medium": "15px",
        "normal": "10px",
        "small": "5px",
        "standard": "20px",
        "xl": "30px",
        "xs": "3px",
        "xxl": "40px",
      },
      "shadows": Object {
        "text": Object {
          "normal": "1px 1px 2px #000",
        },
      },
    }
  }
>
  <App
    componentId="App"
  />
</ThemeProvider>
`;
